internal/encoding/asn1/asn1.go

```diff
@@ -0,0 +1,232 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package asn1 decodes BER-encoded ASN.1 data structures and encodes in DER.
+// Note: DER is a subset of BER.
+// Reference: http://luca.ntop.org/Teaching/Appunti/asn1.html
+package asn1
+
+import (
+	"bytes"
+	"encoding/asn1"
+)
+
+// Common errors
+var (
+	ErrEarlyEOF                    = asn1.SyntaxError{Msg: "early EOF"}
+	ErrTrailingData                = asn1.SyntaxError{Msg: "trailing data"}
+	ErrUnsupportedLength           = asn1.StructuralError{Msg: "length method not supported"}
+	ErrUnsupportedIndefiniteLength = asn1.StructuralError{Msg: "indefinite length not supported"}
+	ErrUnsupportedValueType        = asn1.StructuralError{Msg: "value type not supported"}
+)
+
+// value represents an ASN.1 value.
+type value interface {
+	// Encode encodes the value to the value writer in DER.
+	Encode(*bytes.Buffer) error
+
+	// EncodedLen returns the length in bytes of the encoded data.
+	EncodedLen() int
+}
+
+// ConvertToDER converts BER-encoded ASN.1 data structures to DER-encoded.
+func ConvertToDER(ber []byte) ([]byte, error) {
+	flatValues, err := decode(ber)
+	if err != nil {
+		return nil, err
+	}
+
+	// get the total length from the root value and allocate a buffer
+	buf := bytes.NewBuffer(make([]byte, 0, flatValues[0].EncodedLen()))
+	for _, v := range flatValues {
+		if err = v.Encode(buf); err != nil {
+			return nil, err
+		}
+	}
+
+	return buf.Bytes(), nil
+}
+
+// decode decodes BER-encoded ASN.1 data structures.
+//
+// r is the input byte slice.
+// The returned value is the flat ASN.1 values slice.
+func decode(r []byte) ([]value, error) {
+	// prepare the first value
+	identifier, content, r, err := decodeMetadata(r)
+	if err != nil {
+		return nil, err
+	}
+	if len(r) != 0 {
+		return nil, ErrTrailingData
+	}
+
+	// primitive value
+	if isPrimitive(identifier) {
+		return []value{&primitiveValue{
+			identifier: identifier,
+			content:    content,
+		}}, nil
+	}
+
+	// constructed value
+	rootConstructed := &constructedValue{
+		identifier: identifier,
+		rawContent: content,
+	}
+	flatValues := []value{rootConstructed}
```

Comment:
I want to avoid using type assertion because of the performance overhead. By constructing the `flatValues` during the decoding phase, we can directly obtain the sequential values without needing to perform type assertions to check whether they are primitive or constructed values.